stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -1087,13 +1087,21 @@ const RealOrComplex = Union{Real,Complex}
 # Three-argument *
 """
     *(A, B::AbstractMatrix, C)
+    A * B * C * D
 
-Most 3- and 4-argument `*` calls containing matrices or vectors are done in an
-efficient way, rather than the left-to-right default of `*(x,y,z...)`.
+Chained multiplication of 3 or 4 matrices is done in the most efficient order,
+i.e. the order which minimises the total number of scalar multiplications involved,
+based on the sizes of the arrays.
 
-This can mean performing `B*C` first if `C::AbstractVector`, using five-argument
-`mul!` to fuse the scalar `A::Number` with the matrix multiplication `B*C`,
-or examining `size.((A,B,C,D))` to choose which to multiply first.
+If the last factor is a vector, or the first a transposed vector,
+it is efficient to do these first. In particular `x' * B * y` means `(x' * B) * y`
+for an ordinary colum-major `B`. This is usually faster than `dot(x, B, y)`,
+although it allocates an intermediate array.
```

Comment:
As I'm the "author" of the 3-arg `dot`, I'm a bit sensitive to such statements. 😄 But seriously, I don't think we should conflate the recursive `dot` and the "single-layer" multiplication in the docstring. For the same reason, the term "scalar multiplications" is only correct for scalar eltypes, but not for matrices of matrices. Maybe we should rephrase that to "component/scalar multiplication" or something. Not sure.

NB: As for the speed of 3-arg `dot`, your concerns are valid for the dense matrix case (including symmetric and Hermitian for large sizes). For pretty much every structured/sparse matrix type, I think that 3-arg `dot` should outperform 3-arg `*`.

Reply:
Heh. But sure, the goal I think should be to explain that this may be another option, which is sufficiently different that `x'*A*y` is not automatically sent there.

The whole calculation here is really for dense arrays. If `A,B,C` are sparse matrices, then I guess the fastest sequence will depend a lot on the details? Have not tried to time any.

Reply:
That's all fine. If one only has the docstring, one may miss implicit assumptions, and making "general" recommendations based on hidden assumptions may lead to confusion. I think the general reference to `dot` that you have already given should be fine, and performance-aware users will need to benchmark and compare for their use case.